presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -979,7 +979,16 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
                     throw new SemanticException(INVALID_PROCEDURE_ARGUMENTS, initialValueArg, "REDUCE_AGG only supports non-NULL literal as the initial value", initialValueArg);
                 }
             }
+
             Type type = functionAndTypeManager.getType(functionMetadata.getReturnType());
+
+            if (StandardTypes.MAP.equals(functionMetadata.getReturnType().getBase())) {
+                String keyTypeBase = functionMetadata.getReturnType().getTypeParametersAsTypeSignatures().get(0).getBase();
+                if (StandardTypes.DECIMAL.equals(keyTypeBase) || StandardTypes.DOUBLE.equals(keyTypeBase) || StandardTypes.REAL.equals(keyTypeBase)) {
+                    warningCollector.add(new PrestoWarning(StandardWarningCode.SEMANTIC_WARNING, format("This query uses double as map key which can have non-deterministic results. Node: %s", node)));
```

Comment:
Instead of `Node:` in the warning, I would say `Expression:` since that's more comprehensible to users